sources/engine/Xenko.Graphics/Rendering/EffectInstance.cs

```diff
@@ -92,11 +92,11 @@ protected virtual void ChooseEffect(GraphicsDevice graphicsDevice)
         {
         }
 
-        public void Apply(GraphicsContext graphicsContext)
+        public void Apply(GraphicsContext graphicsContext, int uavInitialOffset = -1)
```

Comment:
The offset is per UAV so it shouldn't end up being global to the effect instance call.
It sounds like it should fit as a `(GraphicsResource, int)` in the `ParameterCollection`.
The issue is that `ParameterCollection` is good to hold either blittable struct (int, float, matrix, etc...) or objects (i.e. GraphicsResource), but not managed struct containing both of them...
We could temporarily wrap this pair of parameter into a class but this is not great for locality/perf (we can deal with that later though).

Reply:
You can still set them one by one, so they should probably be stored as `(GraphicsResource, int)` in the `ParameterCollection` (instead of `GraphicsResource` currently), so that you have all the data to call `descriptorSet.SetUnorderedAccessView` in `EffectParameterUpdater.Update`.

Reply:
had some time to look at this again and it seems that adding the counter value to the `ParameterCollection` is a bit tricky because then the shader key generator (visual studio plugin) would need to generate keys with different types for counter or append buffers. the easiest solution seems to be to add a public `int` field to the `Buffer` type and init it with -1. the only drawback i see would be that the parameter setter would have a try cast to `Buffer` to retrieve the counter value, but that would only be in `SetUnorderedAccessView`;
```csharp
public void SetUnorderedAccessView(int slot, GraphicsResource unorderedAccessView, int uavInitialOffset)
{
    if(unorderedAccessView is Buffer buffer)
    {
        HeapObjects[DescriptorStartOffset + slot].Value = buffer;
        HeapObjects[DescriptorStartOffset + slot].Offset = buffer.UAVInitialOffset;
    }
    else
    {
        HeapObjects[DescriptorStartOffset + slot].Value = unorderedAccessView;
        HeapObjects[DescriptorStartOffset + slot].Offset = -1;
    }
}
```
and to set the buffer:
```csharp
myBuffer.UAVInitialOffset = 0;
myEffect.Parameters.Set(myEffectKeys.MyBuffer, myBuffer);
```

Reply:
Two questions then:
1. Is there some cases where we might want to bind the same buffer with different counts within the same frame?
2. If yes, would this count be kind of static (i.e. -1, 0 or 12 in which case we could allow creating static views) or dynamic (can change any frame and needs to be computed and be updated every draw call within same frame -- in which case your approach would be problematic).

I will also try to look more advanced examples (simple ones seems to always be 0 or -1) or try to think of few scenarios on my side to better understand use cases.

Reply:
yes, the count can change every frame. for example we have a use case where we made something like a RingBuffer data structure with a counter buffer of 10.000 elements divided into 10 sections of 1.000 elements (Dispatch call with max. 1.000 threads) and in each frame we write into one of the sections, then the counter cycles thru values of 0, 1000, 2000 etc. during the frames. the next shader in the pipeline that reads the counter buffer doesn't need to set the counter value, it reads the sections by index.

so is it slow to call `myEffect.Parameters.Set(myEffectKeys.MyBuffer, myBuffer);` each frame? 

but yes this is an advanced case, in mostly it is just -1 or 0.

Reply:
i've edited the comment, the question is: would it be enough to only do `myBuffer.UAVInitialOffset = value;` every frame and the `myEffect.Apply` will take care of the updated value? when the value is a property of the buffer.